data_prep/generate_climos.py

```diff
@@ -1,85 +1,40 @@
 import os, os.path
-import re
 import logging
 import sys
+from itertools import chain
 
 from argparse import ArgumentParser
 from datetime import datetime
-from tempfile import NamedTemporaryFile
 
 from cdo import Cdo
-from netCDF4 import Dataset, num2date, date2num
+from netCDF4 import date2num
 from dateutil.relativedelta import relativedelta
 
-from Cmip5File import Cmip5File, ClimdexFile
+from nchelpers import CFDataset, standard_climo_periods
+from nchelpers.date_utils import d2s
 
-def s2d(s):
-    return datetime.strptime(s, '%Y-%m-%d')
 
-def ss2d(s):
-    return datetime.strptime(s, '%Y%m%d')
+formatter = logging.Formatter('%(asctime)s %(levelname)s: %(message)s', "%Y-%m-%d %H:%M:%S")
+handler = logging.StreamHandler()
+handler.setFormatter(formatter)
 
-def d2s(d):
-    return datetime.strftime(d, '%Y-%m-%d')
+logger = logging.getLogger(__name__)
+logger.addHandler(handler)
 
-def d2ss(d):
-    return datetime.strftime(d, '%Y%m%d')
 
-log = logging.getLogger(__name__)
-logging.basicConfig(stream=sys.stdout, level=4)
+def create_climo_file(outdir, input_file, t_start, t_end):
+    """Generate climatological files from an input file and a selected time range.
 
-climo_periods = {
-    '6190': [s2d('1961-01-01'),s2d('1990-12-31')],
-    '7100': [s2d('1971-01-01'),s2d('2000-12-31')],
-    '8110': [s2d('1981-01-01'),s2d('2010-12-31')],
-    '2020': [s2d('2010-01-01'),s2d('2039-12-31')],
-    '2050': [s2d('2040-01-01'),s2d('2069-12-31')],
-    '2080': [s2d('2070-01-01'),s2d('2099-12-31')]
-}
+    Parameters:
+        outdir (str): path to base directory in which to store output climo file
+        input_file (nchelpers.CFDataset): the input data file
+        t_start (datetime.datetime): start date of climo period to process
+        t_end (datetime.datetime): end date of climo period to process
 
-def iter_matching(dirpath, regexp):
-    # http://stackoverflow.com/questions/4639506/os-walk-with-regex
-    """Generator yielding all files under `dirpath` whose absolute path
-       matches the regular expression `regexp`.
-       Usage:
+    Requested date range MUST exist in the input file.
```

Comment:
Not entirely sure what happens if it doesn't. (This comment was in original script.) What I think happens is that the output file contains only a subset of that date range, and therefore the temporal part of the output filename would be incorrect. First thing that happens in processing is cdo.seldate, which selects a temporal subset of the data in the input file.